jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -330,15 +322,15 @@ public Runnable onIdleTimeout(TimeoutException t)
                 LOG.debug("onIdleTimeout {}", this, t);
 
             // if not already a failure,
-            if (_failure == null)
+            if (_readFailure == null)
```

Comment:
There is a big assumption here (probably correct), that the `_writeFailure` can never be null when the `_readFailure` is not null.  If that was possible, then a new idle timeout would not be delivered to a write because the read had already failed (at least transitively).

I think we should have some asserts to declare these invariants: 

```suggestion
            assert _readFailure == null  || _readFailure != null && _response._writeFailure != null;
            assert _writeFailure == null || _writeFailure != null && _writeFailure.isLast();
            
            if (_readFailure == null)
```

Maybe do the same asserts in onFailure below.

Reply:
@gregw I think there should not be that assumption, and I removed it from the latest code.
The 2 failures should be independent, so I did not add asserts.

Reply:
OK, but you now need to protect the write failure.  See below....